tests/data/apply_filters.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SzepeViktor\PHPStan\WordPress\Tests;
+
+use function apply_filters;
+use function PHPStan\Testing\assertType;
+
+/**
+ * Returns the passed value.
+ *
+ * @template T
+ * @param T $value Value.
+ * @return T Value.
+ */
+function return_value( $value ) {
+    return $value;
+}
+
+$value = apply_filters('filter','Hello, World');
+assertType('mixed', $value);
+
+/**
+ * Unknown parameter.
+ */
+$value = apply_filters('filter',$foo);
+assertType('mixed', $value);
+
+/** @var int $value */
+$value = apply_filters('filter',$foo);
+assertType('int', $value);
+
+/**
+ * Single type.
+ *
+ * @param string $foo Hello, World.
+ */
+$value = apply_filters('filter',$foo);
+assertType('string', $value);
+
+/**
+ * Multiple lines and multiple parameters.
+ *
+ * @param string $foo Hello.
+ * @param bool   $bar World.
+ */
+$value = apply_filters(
+    'filter',
+    $foo,
+    $bar
+);
+assertType('string', $value);
+
+/**
+ * Single constant type.
+ *
+ * @param string $foo Hello, World.
+ */
+$value = apply_filters('filter','I am a string');
+assertType('string', $value);
```

Comment:
@johnbillion I think PHP language is not prepared for this: typehinting a string :)

Could your code handle this? `ConstantStringType` in PHPStan.

Reply:
The return type isn't necessarily the same constant string though, otherwise `apply_filters()` would be a bit useless. Is that what you mean?

Reply:
> The return type

AFAIK you cannot set the return type of literal/constant values with `@param`. So we have to get the type from the constant thus the second parameter.

Reply:
The fact that the parameter is constant doesn't matter. After passing through `apply_filters()` its type is set to whatever the `@param` tag says it is.

```php
add_filter( 'foo', fn() => 456 );

/**
 * @param int $number
 */
$number = apply_filters( 'foo', 123 );
```

The type of `$number` here is `int` not `123`.

Reply:
I think php/PHPStan/any reflection package won't connect `@param int $number` to `123`.
What do you think?

Reply:
That's correct. That's why I've added the dynamic return type extension for `apply_filters()` which reads and "trusts" this WordPress-specific docblock placement.

I would like to write a custom rule at some point which validates the docblock by checking the types for each parameter match what's written in each `@param` tag, but that can be done separately from this dynamic return type extension.

Reply:
I see! `$params[0]->type`